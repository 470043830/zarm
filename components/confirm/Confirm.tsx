import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
import PropsType from './PropsType';
import Modal from '../modal';
import confirmLocale from './locale/zh_CN';
import { getRunTimeLocale } from '../locale-provider/LocaleProvider';

export interface ConfirmProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Confirm extends PureComponent<ConfirmProps, {}> {
  static defaultProps = {
    prefixCls: 'za-confirm',
    animationType: 'zoom',
    locale: confirmLocale,
  };

  private static confirmContainer;

  static show = (props: ConfirmProps) => {
    const { defaultProps } = Confirm;
    let _props: any;

    if (typeof props === 'object') {
      _props = { ...defaultProps, ...props };
    } else {
      _props = { ...defaultProps };
    }

    const runTimeLocale = getRunTimeLocale();
    if (runTimeLocale && runTimeLocale.Confirm) {
      _props = { ..._props, locale: runTimeLocale.Confirm };
    }
    Confirm.confirmContainer = document.createElement('div');
    document.body.appendChild(Confirm.confirmContainer);
    return new Promise((resolve) => {
      ReactDOM.render(
        <Confirm
          {..._props}
          visible
          onOk={() => {
            if (typeof _props.onOk === 'function') {
              _props.onOk();
            }
            resolve(true);
          }}
          onCancel={() => {
            if (typeof _props.onCancel === 'function') {
              _props.onCancel();
            }
            resolve(false);
          }}
        />,
        Confirm.confirmContainer,
      );
    });
  };

  static hide = () => {
    if (Confirm._hide) {
      Confirm._hide();
    }
  };

  static _hide: () => void;

  state = {
    visible: this.props.visible,
  };

  componentDidMount() {
    Confirm._hide = this._hide;
  }

  static getDerivedStateFromProps(nextProps, state) {
    if ('visible' in nextProps && nextProps.visible !== state.prevVisible) {
      return {
        visible: nextProps.visible,
        prevVisible: nextProps.visible,
      };
    }
    return null;
  }

  _hide = () => {
    this.setState({
      visible: false,
    });
  };

  afterClose = () => {
    const { afterClose } = this.props;
    if (Confirm.confirmContainer) {
      document.body.removeChild(Confirm.confirmContainer);
      Confirm.confirmContainer = null;
    }

    if (typeof afterClose === 'function') {
      afterClose();
    }
  };

  render() {
    const { prefixCls, className, message, okText, cancelText, onOk, onCancel, locale, ...others } = this.props;
    const { visible } = this.state;
    const cls = classnames(prefixCls, className);
    return (
      <Modal
        className={cls}
        {...others}
        visible={visible}
        afterClose={this.afterClose}
        footer={(
          <>
            <div className={`${prefixCls}__button`} onClick={onCancel}>{cancelText || locale!.cancelText}</div>
            <div className={`${prefixCls}__button ${prefixCls}__button--ok`} onClick={onOk}>
              {okText || locale!.okText}
            </div>
          </>
        )}
      >
        {message}
      </Modal>
    );
  }
}
