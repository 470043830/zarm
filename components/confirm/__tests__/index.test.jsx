import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Confirm from '../index';

describe('Confirm', () => {
  it('renders correctly', () => {
    const onOk = jest.fn();
    const onCancel = jest.fn();

    const wrapper = render(
      <Confirm
        radius
        visible
        title="确认信息"
        message="你确定要这样做吗？"
        onOk={onOk}
        onCancel={onCancel}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('works with static methods', () => {
    const onCancel = jest.fn();
    const promise = Confirm.show({
      title: '警告',
      message: '这是警告信息',
      onCancel,
    });
    document.querySelector('.za-confirm__button').click();

    expect(onCancel).toBeCalledTimes(1);

    return expect(promise).resolves.toBe(false);
  });
});
