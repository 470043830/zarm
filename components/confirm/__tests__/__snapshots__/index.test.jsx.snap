// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm renders correctly 1`] = `
<div
  class="za-modal za-confirm za-modal--radius"
  style="-webkit-animation-duration:200ms;animation-duration:200ms"
>
  <div
    class="za-modal__wrapper"
  >
    <div
      class="za-modal__dialog"
      style="width:70%;-webkit-animation-duration:200ms;animation-duration:200ms"
    >
      <div
        class="za-modal__header"
      >
        <div
          class="za-modal__header__title"
        >
          确认信息
        </div>
      </div>
      <div
        class="za-modal__body za-confirm__body"
      >
        你确定要这样做吗？
      </div>
      <div
        class="za-modal__footer za-confirm__footer"
      >
        <div
          class="za-confirm__footer__button"
        >
          Cancel
        </div>
        <div
          class="za-confirm__footer__button za-confirm__footer__button--ok"
        >
          OK
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-mask za-mask--normal"
  />
</div>
`;
